src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py

```diff
@@ -0,0 +1,187 @@
+            sys.exit(0)
+        # get chosen arch from tuner
+        self.chosen_arch = nni.get_next_parameter()
+        if not self.chosen_arch and trial_env_vars.NNI_PLATFORM is None:
+            logger.warning('This is in standalone mode, the chosen are the first one(s)')
+            self.chosen_arch = self._standalone_generate_chosen()
+        self._validate_chosen_arch()
+
+    def _validate_chosen_arch(self):
+        pass
+
+    def _standalone_generate_chosen(self):
+        """
+        Generate the chosen architecture for standalone mode,
+        i.e., choose the first one(s) for LayerChoice and InputChoice
+
+        Returns
+        -------
+        dict
+            the chosen architecture
+        """
+        chosen_arch = {}
+        for key, val in self.search_space.items():
+            if val['_type'] == 'layer_choice':
+                choices = val['_value']
+                chosen_arch[key] = choices[0]
+            elif val['_type'] == 'input_choice':
+                choices = val['_value']['candidates']
+                n_chosen = val['_value']['n_chosen']
+                chosen = []
+                for i in range(n_chosen):
+                    chosen.append(choices[i])
+                chosen_arch[key] = chosen
+            else:
+                raise ValueError('Unknown key %s and value %s' % (key, val))
+        return chosen_arch
+
+    def _generate_search_space(self):
+        """
+        Generate search space from mutables.
+        Here is the search space format:
+
+        { key_name: {'_type': 'layer_choice',
+                     '_value': [conv1, conv2]} }
+
+        { key_name: {'_type': 'input_choice',
+                     '_value': {'candidates': [in1, in2],
+                                'n_chosen': 1}} }
+
+        Returns
+        -------
+        dict
+            the generated search space
+        """
+        search_space = {}
+        for mutable in self.mutables:
+            # for now we only generate flattened search space
+            if isinstance(mutable, LayerChoice):
+                key = mutable.key
+                val = [choice.__repr__() for choice in mutable.choices]
+                search_space[key] = {"_type": "layer_choice", "_value": val}
+            elif isinstance(mutable, InputChoice):
+                key = mutable.key
+                search_space[key] = {"_type": "input_choice",
+                                     "_value": {"candidates": mutable.choose_from,
+                                                "n_chosen": mutable.n_chosen}}
+            else:
+                raise TypeError('Unsupported mutable type.')
+        return search_space
+
+    def _dump_search_space(self, search_space):
+        with open(os.path.join(os.getcwd(), 'auto_gen_search_space.json'), 'w') as ss_file:
+            json.dump(search_space, ss_file)
+
+    def _tensor_reduction(self, reduction_type, tensor_list):
+        if tensor_list == "none":
```

Comment:
Duplicated code snippets. Needs refactor.

Reply:
will refactor in future